docs/design/2021-09-22-data-consistency.md

```diff
@@ -0,0 +1,121 @@
+- MustExist(10): Key must exist in storage, subsequent modifications within the transaction cannot be overwritten
+- MustNotExist(01): Key must not exist in the storage, subsequent modifications within the transaction cannot be overwritten
+
+The flags in MemDB, like the KV data, are first set to the current stage of the statement. After that, if the statement reports an error and the statement rolls back, the flags will be invalidated. If the statement is executed successfully and is submitted, the flags will take effect in the transaction. The assertion information is sent to TiKV in the prewrite stage of 2PC. TiKV performs additional checks based on the assertions before modifying data. If the assertions fail, TiKV rejects the request.
+
+### Check Mutations
+
+The basic idea is to ensure that the transaction will not write inconsistent data when the data before the transaction is consistent.
+
+For possible errors in the executor: Assume that the set of row values before and after the transaction are V1 and V2 respectively, and the data indices are consistent before the transaction is executed. The condition of data consistency after the transaction's commitment is V1-V2 = { Del_{index}.value }, V2-V1 = { Put_{index}.value }. 
+
+## Test Design
+
+The main quality objectives of this project include correctness, effectiveness, efficiency, ease of use, and compatibility. The test part involves two parts: ensuring correctness (including compatibility) and measuring effectiveness. 
+
+Correctness: There will be no false positives when the data is correct 
+
+Effectiveness: 
+
+1. Reject transactions with inconsistent mutations 
+2. Report errors as soon as possible for existing problematic data to prevent the spread of errors 
+
+### Regression tests
+
+Regression tests are useful to check correctness since all tests should not report data inconsistency errors.
+
+### System tests
+
+Specific system test cases need to be constructed. Cases should at least involve the following parts
+
+- encoding related
+
+- - collation
+  - row format
+  - charset
+  - time zone
+
+- table structure
+
+- - concurrent DDL
+  - clustered index
+  - prefix index
+  - expression index
+
+### Performance Tests
+
+The enhancement should have little impact on performance, including latency, throughput, memory and CPU consumption.
+
+Standard sysbench and TPC-C tests are needed. 
+
+## Impacts & Risks
+
+In the assertion plan, pessimistic transactions may need an extra read in the WRITE CF when prewriting. 
+
+When the CPU becomes the bottleneck of TiDB, introducing the enhancement may cause performance degradation.
+
+Further benchmark is needed to determine the precise influence of the risks.
+
+## Unresolved Questions
+
+The way to treat pessimistic transactions in the assertion plan can be modified, or redesigned if the performance influence is notable.
```

Comment:
It seems this needs to be updated. How about adding some description about we checking assertion according to existence check result of pessimistic lock requests?